src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

```diff
@@ -157,19 +226,40 @@ class GoogleKeyCache(
   private def keyNameFull(project: GoogleProject, saEmail: WorkbenchEmail, keyId: ServiceAccountKeyId): GcsBlobName =
     GcsBlobName(s"${keyNamePrefix(project, saEmail)}${keyId.value}")
 
+  /** Cleans up keys by deleting orphans and bringing total key count down to 9, then creates a new key.
+    */
+  private def cleanupAndCreateKey(pet: PetServiceAccount, cachedKeyObjects: List[CachedKey]): IO[String] =
+    for {
+      keysInIam <- fetchKeysFromIam(pet)
+      _ <- cleanupKeys(pet, cachedKeyObjects, keysInIam)
+      newKey <- furnishNewKey(pet)
+    } yield newKey
+
+  /** Create a new key for this pet. Do not call this directly, as it has no protection against running over the 10-key limit. Call cleanupAndCreateKey()
+    * instead.
+    */
   private def furnishNewKey(pet: PetServiceAccount): IO[String] =
     for {
       key <- IO.fromFuture(IO(googleIamDAO.createServiceAccountKey(pet.id.project, pet.serviceAccount.email))) recover {
-        // TODO CORE-278: on error, check the number of existing keys and purge as necessary
         case e: GoogleJsonResponseException =>
+          logger.error(
+            s"furnishNewKey: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} ${e.getMessage}",
+            e
+          )
           if (e.getDetails.getCode == StatusCodes.TooManyRequests.intValue)
-            // 2025-01-29: TooManyRequests is not returned by Google for this error any more. Leaving this in place
+            // 2025-01-29: TooManyRequests is not returned by Google for this error anymore. Leaving this in place
             //  in case Google switches back to it
             throw new WorkbenchException("You have reached the 10 key limit on service accounts. Please remove one to create another.")
           else if (e.getDetails.getCode == StatusCodes.BadRequest.intValue && e.getDetails.getMessage == "Precondition check failed.")
             throw new WorkbenchException("You may have reached the 10 key limit on service accounts.")
           else
             throw new WorkbenchException(s"Error creating key for service account: ${e.getDetails.getCode}: ${e.getDetails.getMessage}")
+        case t: Throwable =>
+          logger.error(
+            s"furnishNewKey: ${pet.id.project.value}-${pet.serviceAccount.subjectId.value} ${t.getMessage}",
+            t
+          )
+          throw new WorkbenchException(s"Error creating key for service account: ${t.getMessage}")
```

Comment:
this is the antipattern of both logging and throwing. I really wanted to log the stack trace for these errors, but really didn't want to include the stack trace when showing the exception to the user … so I'm logging and throwing.